crates/sui-bcs-json/src/lib.rs

```diff
@@ -0,0 +1,829 @@
+        }
+    };
+
+    match internal_bcs_from_json(type_name, json_data) {
+        Ok(output) => {
+            let mut vec = output;
+            vec.shrink_to_fit();
+            assert!(vec.len() == vec.capacity());
+            let ptr = vec.as_mut_ptr();
+            let len = vec.len();
+            // Prevent running `vec`'s destructor so we are in complete control
+            // of the allocation.
+            let _ = ManuallyDrop::new(vec);
+            unsafe {
+                *result = ptr;
+                *length = len;
+            }
+            0
+        }
+        Err(e) => {
+            ffi_helpers::update_last_error(e);
+            1
+        }
+    }
+}
+
+/// Return 0 if the conversion from BCS to JSON is successful, and 1 or 2 for
+/// failure. 1 represents a failure from parsing the BCS to JSON, and 2
+/// represents an error building the CString from the JSON data.
+///
+/// The result argument will point to the address where the JSON
+/// string is stored. Make sure you release the allocated memory
+/// by calling sui_bcs_json_free_string function!
+///
+/// # Safety
+/// Unsafe function.
+#[no_mangle]
+pub unsafe extern "C" fn sui_bcs_to_json(
+    type_name: *const c_char,
+    bcs_ptr: *const u8,
+    len: usize,
+    result: *mut *const c_char,
+    pretty: bool,
+) -> usize {
+    let type_name = match unsafe { CStr::from_ptr(type_name) }.to_str() {
+        Ok(type_name) => type_name,
+        Err(e) => {
+            ffi_helpers::update_last_error(e);
+            return 1;
+        }
+    };
+
+    // we have a pointer to a pointer, thus we first get the address of the pointer
+    // to the bcs array data and then reconstruct the slice
+    let mut data = unsafe { std::slice::from_raw_parts(bcs_ptr, len) };
+    match internal_bcs_to_json(type_name, &mut data, pretty) {
+        Ok(res) => {
+            let cstr = std::ffi::CString::new(res);
+            let cstr = match cstr {
+                Ok(c) => c.into_raw(),
+                Err(e) => {
+                    ffi_helpers::update_last_error(e);
+                    return 2;
+                }
+            };
+            unsafe { *result = cstr as *const i8 };
+            0
+        }
+        Err(e) => {
+            ffi_helpers::update_last_error(e);
+            1
+        }
+    }
+}
+
+/// Frees a Rust-allocated `Vec<u8>`.
+#[no_mangle]
+pub extern "C" fn sui_bcs_json_free_array(ptr: *const u8, len: usize) {
+    unsafe {
+        // SAFETY: not going to mutate the value via the pointer, so this is fine
+        drop(Vec::from_raw_parts(ptr as *mut u8, len, len));
+    }
+}
+
+/// Frees a Rust-allocated string.
+#[no_mangle]
+pub extern "C" fn sui_bcs_json_free_string(pointer: *const c_char) {
+    let _ = unsafe { std::ffi::CString::from_raw(pointer as *mut i8) };
+}
+
+/// Deserializes the data to given type T, and then serializes it as a JSON value
+fn to_json<T: Serialize + for<'a> Deserialize<'a>>(data: &mut &[u8]) -> Result<Value> {
```

Comment:
Do you still need the existential lifetime qualification, or can you get rid of it now?  Also, can you get rid of the `&mut` now?

```suggestion
fn to_json<T: Serialize + Deserialize>(data: &[u8]) -> Result<Value> {
```

Reply:
If we go with the updated ser/de, these do not make sense anymore.